src/fetch.rs

```diff
@@ -0,0 +1,362 @@
+use std::{
+    fmt::{self, Debug},
+    path::PathBuf,
+    str::FromStr,
+};
+
+use anyhow::{anyhow, Context, Result};
+use bytes::Bytes;
+use reqwest::{
+    blocking::Client,
+    header::{ETAG, IF_NONE_MATCH, USER_AGENT},
+    StatusCode,
+};
+
+use rpki::{
+    rrdp::{Delta, DeltaInfo, Hash, NotificationFile, Snapshot, SnapshotInfo},
+    uri,
+};
+
+use crate::{config, file_ops};
+
+//------------ FetchResponse -------------------------------------------------
+enum FetchResponse {
+    Data { bytes: Bytes, etag: Option<String> },
+    UnModified,
+}
+
+//------------ NotificationFileResponse --------------------------------------
+pub enum NotificationFileResponse {
+    Data {
+        notification: NotificationFile,
+        etag: Option<String>,
+    },
+    Unmodified,
+}
+
+impl NotificationFileResponse {
+    pub fn content(self) -> Result<(NotificationFile, Option<String>)> {
+        match self {
+            NotificationFileResponse::Data { notification, etag } => Ok((notification, etag)),
+            NotificationFileResponse::Unmodified => {
+                Err(anyhow!("Cannot get content from unmodified response"))
+            }
+        }
+    }
+}
+
+//------------ FetchSource ---------------------------------------------------
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum FetchSource {
+    File(PathBuf),
+    Uri(uri::Https),
+}
+
+impl FetchSource {
+    /// Gets the data from the fetch source, and verifies that it matches
+    /// the hash - if it is provided.
+    fn fetch(&self, hash: Option<Hash>, etag: Option<&String>) -> Result<FetchResponse> {
+        let fetch_response = match self {
+            FetchSource::Uri(uri) => {
+                let mut request_builder = Client::builder().build()?.get(uri.as_str());
```

Comment:
A couple of observations here:
- Previously a HTTP client was created once and used for all connections. This might have meant it was able to re-use pooled connections perhaps even without redoing TLS setup. I suspect, but haven't yet verified, that this new code calls `fetch()` multiple times each time creating a new client thereby losing these optimisations.
- Previously the HTTP client created by Routinator had timeouts set, this code doesn't seem to set timeouts. E.g. see https://github.com/ximon18/routinator/blob/31b37bd1f07e7c2bb156e4f517888dd4bce9cc13/src/rrdp/http.rs#L64.

Reply:
Remember that this is (mostly?) intended to run on a local network between the actual backend publication server and the front-ends: 'public' https/rsyncd.

Apart from the initial bootstrap we should expect only one or a few deltas for each synchronisation. So, client re-use was not done as it would add complexity to the design.

The default timeout is used for now, which is 30 seconds. We can easily change it or add a config parameter for it if desired. But I believe that this should normally be fine.

So, my take on this in a nutshell is that we should test this with the people using this first - and if needed we can add the complexity and optimisations.